subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/TestLauncherDebugCrossVersionTest.groovy

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.tooling.r85
+
+import org.apache.commons.io.output.TeeOutputStream
+import org.gradle.integtests.tooling.fixture.TargetGradleVersion
+import org.gradle.integtests.tooling.fixture.ToolingApiSpecification
+import org.gradle.test.fixtures.file.TestFile
+import org.gradle.tooling.GradleConnectionException
+import org.gradle.tooling.ProjectConnection
+import org.gradle.tooling.ResultHandler
+import spock.lang.Issue
+import spock.lang.Retry
+
+@Retry(count = 3)
+@TargetGradleVersion('>=8.5')
+class TestLauncherDebugCrossVersionTest extends ToolingApiSpecification {
+
+    @Issue('')
+    def "Run test twice (#scenarioName)"() {
+        setup:
+        sampleBuildWithTest()
+
+        when:
+        String output1 = runTaskAndTestClassUsing(firstDebug)
+        String output2 = runTaskAndTestClassUsing(secondDebug)
+
+        then:
+        assertTestDebugMode(output1, firstDebug)
+        assertTestDebugMode(output2, secondDebug)
+
+        where:
+        scenarioName                     | firstDebug | secondDebug
+        "first wo/debug, second w/debug" | false      | true
+        "first w/debug, second wo/debug" | true       | false
+    }
+
+    private def sampleBuildWithTest() {
+        settingsFile << "include('app')"
+        javaBuildWithTests(file('app'))
+    }
+
+    private String runTaskAndTestClassUsing(boolean debugMode) {
+        def stdout = new ByteArrayOutputStream()
+        withConnection { ProjectConnection connection ->
+            def tee = new TeeOutputStream(System.out, stdout)
+            def testLauncher = connection.newTestLauncher()
+                .withTaskAndTestClasses(':app:test', Arrays.asList('TestClass1'))
+                .setStandardOutput(tee)
+            if (debugMode) {
+                // JDWPUtil is flaky https://github.com/gradle/gradle/issues/26366
+                testLauncher.debugTestsOn(49052 + new Random(System.currentTimeMillis()).nextInt(100))
```

Comment:
❗ This makes the test non-deterministic, harder to debug and more likely to become flaky.  Is a static seed possible, perhaps incrementing until a free port is found?

Reply:
This is a workaround until https://github.com/gradle/gradle-private/issues/3986 is fixed (need to adjust the URL there). There's something funky about detecting free ports on CI, thence the workaround with a random number.

Reply:
Btw this test doesn't allocate any port at all, just needs the debugger to not actively listen on the selected random port. Any other process using that port is fine.

Reply:
Fixed